File: Threads Posix/exo8_b.py
import threading
import random
import time
import sys

# La barrière sera initialisée dans le main
barriere = None

def travail_thread(thread_id):
    # Simulation d'un travail de durée aléatoire
    time.sleep(random.randint(1, 5))
    print(f"Thread {thread_id} arrivé au rendez-vous.")
    
    try:
        # Équivalent de pthread_barrier_wait
        barriere.wait()
        
        print(f"Thread {thread_id} franchit la barrière !")
    except threading.BrokenBarrierError:
        # Gère le cas où la barrière est réinitialisée ou brisée
        print(f"Thread {thread_id} : La barrière a été brisée.")

def main():
    global barriere
    
    if len(sys.argv) < 2:
        return

    try:
        n = int(sys.argv[1])
    except ValueError:
        return

    # Initialisation de la barrière pour n threads
    # Équivalent de pthread_barrier_init(&barriere, NULL, n)
    barriere = threading.Barrier(n)

    threads = []
    
    # Création et lancement des threads
    for i in range(n):
        t = threading.Thread(target=travail_thread, args=(i,))
        threads.append(t)
        t.start()

    # Attente de la fin de tous les threads (pthread_join)
    for t in threads:
        t.join()

    print("Tous les threads ont terminé.")

if __name__ == "__main__":
    main()

File: Threads Posix/exo8_b.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h>


pthread_barrier_t barriere; 

void* travail_thread(void* arg) 
{

    int id = *((int*)arg);
    sleep((rand() % 5) + 1);
    printf("Thread %d arrivé au rendez-vous.\n", id);
    
    // Bloque ici jusqu'à ce que N threads aient appelé cette fonction
    pthread_barrier_wait(&barriere);
    
    printf("Thread %d franchit la barrière !\n", id);
    return NULL;
}

int main(int argc, char *argv[]) 
{
    if (argc < 2) 
    {
        return 1;
    }
    int n = atoi(argv[1]);
    pthread_t threads[n];
    int ids[n];

    pthread_barrier_init(&barriere, NULL, n);

    for (int i = 0; i < n; i++) 
    {
        ids[i] = i;
        pthread_create(&threads[i], NULL, travail_thread, &ids[i]);
    }

    for (int i = 0; i < n; i++) 
    {
        pthread_join(threads[i], NULL);
    }
    pthread_barrier_destroy(&barriere);
    return 0;
}